### ISSUE TITLE
Variable-handling error introduced in 2.1.1.0

### ISSUE THREAD

**Issue by djmitche** [ansible/ansible#17356]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

core
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /home/dustin/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

(issue verified to not occur with 2.1.0.0)
##### CONFIGURATION

```
$ cat ansible.cfg
[defaults]
inventory = inventory
retry_files_enabled = False

$ cat inventory
localhost

$ cat roles/projects/tasks/main.yml

---
- include: "./sub.yml"
  tags: projects
  vars:
    name: "{{item}}"
  with_items: "{{projects}}"

$ cat roles/projects/tasks/sub.yml

---
# removing this debug will cause the playbook to succeed
- debug:
    msg: "hi"
- include: "./subsub.yml"

$ cat roles/projects/tasks/subsub.yml

---
- debug:
    msg: "name is {{ name }}"

$ cat run.yml

---
- name: "Test"
  hosts: all
  connection: local
  gather_facts: yes
  roles:
  - role: projects
    projects:
    - sub
```
##### OS / ENVIRONMENT

Ubuntu
##### SUMMARY

This is a minimal reproduction of an error that is causing my configuration to fail on 2.1.1.0, where it succeeded in 2.1.0.0.  The error message is not especially useful.  The surprising thing about the reproduction is, removing the commented "debug" will cause it to succeed -- so that debug is having some unfortunate side-effect!
##### STEPS TO REPRODUCE

Using the files above:

```
$ ansible-playbook -vvvv run.yml
```

(results in ACTUAL RESULTS, below)
##### EXPECTED RESULTS

With 2.1.0.0:

```
$ ansible-playbook run.yml

PLAY [Test] ********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [projects : include] ******************************************************
included: /home/dustin/ansible/roles/projects/tasks/./sub.yml for localhost

TASK [projects : debug] ********************************************************
ok: [localhost] => {
    "msg": "hi"
}

TASK [projects : include] ******************************************************
included: /home/dustin/ansible/roles/projects/tasks/subsub.yml for localhost

TASK [projects : debug] ********************************************************
ok: [localhost] => {
    "msg": "name is sub"
}

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0
```
##### ACTUAL RESULTS

```
$ ansible-playbook run.yml
Using /home/dustin/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: run.yml **************************************************************
1 plays in run.yml

PLAY [Test] ********************************************************************

TASK [setup] *******************************************************************
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: dustin
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1472762126.82-79200200000012 `" && echo ansible-tmp-1472762126.82-79200200000012="` echo $HOME/.ansible/tmp/ansible-tmp-1472762126.82-79200200000012 `" ) && sleep 0'
<localhost> PUT /tmp/tmpZmbYvM TO /home/dustin/.ansible/tmp/ansible-tmp-1472762126.82-79200200000012/setup
<localhost> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/dustin/.ansible/tmp/ansible-tmp-1472762126.82-79200200000012/setup; rm -rf "/home/dustin/.ansible/tmp/ansible-tmp-1472762126.82-79200200000012/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [projects : include] ******************************************************
task path: /home/dustin/ansible/roles/projects/tasks/main.yml:2
included: /home/dustin/ansible/roles/projects/tasks/./sub.yml for localhost

TASK [projects : debug] ********************************************************
task path: /home/dustin/ansible/roles/projects/tasks/sub.yml:3
ok: [localhost] => {
    "msg": "hi"
}

TASK [projects : debug] ********************************************************
task path: /home/dustin/ansible/roles/projects/tasks/subsub.yml:2
fatal: [localhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: {{item}}: 'item' is undefined\n\nThe error appears to have been in '/home/dustin/ansible/roles/projects/tasks/subsub.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- debug:\n  ^ here\n"}

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1   
```

**Comment by stewartm** [comment]
+1

In my case I have an include task that is passed a loop. The loop_var is set by loop_control and accessible within the included task. However, the include task contains a further include which cannot access the loop_var.

Works fine with 2.1.0.0. Breaks with 2.1.1.0.

Will expand this comment with an example when I have a little more time.

**Comment by jctanner** [comment]
@djmitche thank you for bringing this to our attention. I have confirmed the reported behavior with 2.1.0 and 2.1.1. However, it seems to be working in devel as of https://github.com/ansible/ansible/commits/6717710

We will be pushing a new release very soon after a lockdown of the current codebase, so please update once we release and retest your scenario.

**Comment by djmitche** [comment]
I'm a little curious what went wrong here.  I consider Ansible a programming language, and if Python released 2.7.21 that accidentally broke otherwise-valid code, I don't think "well, seems fine in 2.7.22" would be an adequate resolution.  At the very least, I would expect a regression test to prevent this from accidentally happening again.

git bisect on the main repo shows 73a3a5839b7698a6b2d6ea459b0f65234729627b (#16192) being the bad commit.

438ed70a4373f1035ae90a042bc745791b40fa76 is the fix.  Not surprisingly, it is a revert of 73a3a5839b7698a6b2d6ea459b0f65234729627b.

```
commit 438ed70a4373f1035ae90a042bc745791b40fa76
Author: Martin Matuska <martin.matuska@axelspringer.de>
Date:   Mon May 2 17:50:42 2016 +0200

    Restore Ansible 2.0 compatibility for includes

diff --git a/lib/ansible/playbook/helpers.py b/lib/ansible/playbook/helpers.py
index 608f3ee..5ee979e 100644
--- a/lib/ansible/playbook/helpers.py
+++ b/lib/ansible/playbook/helpers.py
@@ -109,18 +109,17 @@ def load_list_of_tasks(ds, play, block=None, role=None, task_include=None, use_h
                 templar = Templar(loader=loader, variables=all_vars)

                 # check to see if this include is static, which can be true if:
                 # 1. the user set the 'static' option to true
                 # 2. one of the appropriate config options was set
                 # 3. the included file name contains no variables, and has no loop
                 is_static = t.static or \
                             C.DEFAULT_TASK_INCLUDES_STATIC or \
-                            (use_handlers and C.DEFAULT_HANDLER_INCLUDES_STATIC) or \
-                            not templar._contains_vars(t.args.get('_raw_params')) and t.loop is None
+                            (use_handlers and C.DEFAULT_HANDLER_INCLUDES_STATIC)

                 if is_static:
                     if t.loop is not None:
                         raise AnsibleParserError("You cannot use 'static' on an include with a loop", obj=task_ds)

                     # FIXME: all of this code is very similar (if not identical) to that in 
                     #        plugins/strategy/__init__.py, and should be unified to avoid
                     #        patches only being applied to one or the other location
```

#16192 was supposed to fix #15735.  I assume that this revert has re-broken #15735, so perhaps it should be re-opened?

**Comment by djmitche** [comment]
Ah, I see now that #15689 appears to have solved #15735 in a different way.  So at least we have some background beyond the "seems to be fixed".  A regression test would be nice, but I'm sure I'll be back to file another issue if I see this again :)
